### PR TITLE
feat(pnp): exports support

### LINKED ISSUE #650: [Case Study] Exports Support
- [ ] I'd be willing to implement a fix

**Describe the bug**

See:
- [proposal-pkg-exports](https://github.com/jkrems/proposal-pkg-exports)
- [Node 13 Documentation](https://nodejs.org/api/esm.html#esm_package_exports)

[Node ^12.7.0](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V12.md#12.7.0) with the `--experimental-modules` flag and [Node ^13.0.0](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V13.md#13.0.0) without the flag resolves modules differently than Yarn@2.0.0-rc19 with PNP 

**To Reproduce**

See this project:
https://github.com/sparebytes/yarn2-pnp-exports-example

```
// packages/package-with-exports/package.json
{
    "name": "package-with-exports",
    "version": "1.0.0",
    "exports": {
        ".": "./main.js",
        "./sub/other": "./other.js"
    }
}

// Example 1
// Expected: packages/package-with-exports/main.js
// Actual: packages/package-with-exports/index.js
require("package-with-exports");

// Example 2
// Expected: packages/package-with-exports/other.js
// Actual: throws error
require("package-with-exports/sub/other");
```

**Environment**

 - OS: Windows
 - Node version ^12.7.0
 - Yarn version 2.0.0-rc19

### PR DESCRIPTION
**Pull request by paul-soporan**
**What's the problem this PR addresses?**

I'm opening this as a draft so that people know that I'm working on it (and it's nearly finished).

This PR adds support for the "exports" field to PnP (CJS only, @merceyz is looking into ESM in #2161) via the [`resolve.exports`](https://github.com/lukeed/resolve.exports) package.

Closes #650.

Edit: Forgot to give credit earlier, but this PR adapts some of @bgotink's code from https://github.com/yarnpkg/berry/pull/1359.

TODO:
- [x] Add test checking that `resolveUnqualifiedExport` is a method on the pnpapi
- [x] Don't throw if the manifest doesn't exist
- [x] Fix issue with packages that are discarded from lookup (`link:` is currently completely broken)
- [x] Add more tests for the `link:` edge cases
- [ ] Allow passing conditions directly to `resolveRequest` & `resolveUnqualifiedExport` - https://github.com/yarnpkg/berry/pull/2431#discussion_r573756915
- [ ] Use longest matching path - waiting on https://github.com/lukeed/resolve.exports/pull/8
- [ ] Exports validation - waiting on https://github.com/lukeed/resolve.exports/pull/6
- [ ] Support for the `--conditions` flag - waiting on https://github.com/nodejs/node/issues/36935 (will probably happen in another PR)

**Checklist**

- [X] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [X] I have set the packages that need to be released for my changes to be effective.

- [X] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/yarnpkg-pnp/sources/loader/makeApi.ts [hunk 1]
```diff
@@ -172,9 +173,51 @@ export function makeApi(runtimeState: RuntimeState, opts: MakeApiOptions): PnpAp
...
+
+    if (!/^\.{0,2}\//.test(subpath))
+      subpath = `./${subpath}` as PortablePath;
+
+    const resolvedExport = resolveExport(pkgJson, ppath.normalize(subpath), {
+      require: true,
+      browser: false,
+      // TODO: implement support for the --conditions flag
+      // Waiting on https://github.com/nodejs/node/issues/36935
+      conditions: [],
```
Comment (merceyz) [inline packages/yarnpkg-pnp/sources/loader/makeApi.ts:209]: Should be allowed to pass this through `resolveUnqualifiedExport` for API consumers like the ESM loader in #2161

File: packages/yarnpkg-pnp/sources/loader/makeApi.ts [hunk 2]
```diff
@@ -203,9 +192,51 @@ export function makeApi(runtimeState: RuntimeState, opts: MakeApiOptions): PnpAp
...
+    if (!/^\.{0,2}\//.test(subpath))
+      subpath = `./${subpath}` as PortablePath;
+
+    const resolvedExport = resolveExport(pkgJson, ppath.normalize(subpath), {
+      browser: false,
+      require,
+      // TODO: implement support for the --conditions flag
+      // Waiting on https://github.com/nodejs/node/issues/36935
+      conditions,
+    });
```
Comment (merceyz) [inline packages/yarnpkg-pnp/sources/loader/makeApi.ts:229]: I mentioned it on Discord but just to have it here as well so it isn't forgotten:

No matter what we pass here `resolve.exports` [forces two conditions from `['require', 'import', 'browser', 'node']` to always be added](https://github.com/lukeed/resolve.exports/blob/c6814c43cb27ee6092f649be3fdf6c1fb55c51e4/src/index.js#L70-L72) while we need it to only use the conditions we pass so we have full control. For example if `react-native` were to use this to resolve native stuff it should not get matches for neither `node` nor `browser`. While in the ESM loader the conditions can be anything some other loader passes so the same applies there.

cc @lukeed

Comment (lukeed) [reply packages/yarnpkg-pnp/sources/loader/makeApi.ts:236]: That is correct. See [`defaultConditions`](https://nodejs.org/api/esm.html#esm_resolver_algorithm) and also [here](https://nodejs.org/api/packages.html#packages_conditional_exports). Node is always juggling up to 4 conditions (with 2 forced as defaults). 

You can add whatever custom conditions you want to support, but at the end of the day, the specified key order determines the matching priority. A RN user will need to do something like this for their package to work anywhere (not just _my_ module):

```js
{
  "exports": {
    "react-native": {
      "import": "./rn/index.mjs",
      "require": "./rn/index.cjs"
    },
    "import": "./...",
     //...
  }
}
```

And then specify the "react-native" condition

File: packages/yarnpkg-pnp/sources/loader/hydrateRuntimeState.ts [hunk 3]
```diff
@@ -14,29 +14,29 @@ export function hydrateRuntimeState(data: SerializedState, {basePath}: HydrateRu
...
-        // @ts-expect-error: TypeScript isn't smart enough to understand the type assertion
-        const packageLocator: PhysicalPackageLocator = {name: packageName, reference: packageReference};
-        packageLocatorsByLocations.set(packageInformationData.packageLocation, packageLocator);
+      const discardFromLookup = packageInformationData.discardFromLookup ?? false;
 
-        packageLocationLengths.add(packageInformationData.packageLocation.length);
-      }
+      // @ts-expect-error: TypeScript isn't smart enough to understand the type assertion
+      const packageLocator: PhysicalPackageLocator = {name: packageName, reference: packageReference};
+      packageLocatorsByLocations.set(packageInformationData.packageLocation, {locator: packageLocator, discardFromLookup});
```
Comment (arcanis) [inline packages/yarnpkg-pnp/sources/loader/hydrateRuntimeState.ts:29]: The bug comes from here - because of `link:`, you can end up in a state where two "packages" have the same location: all you need is to make a package have a `link:.` dependency (which is what the failing test does). In general it's not a problem, but since you now attach the `discardFromLooking` flag, the `false` value overwrites the `true`, and the path can't be required anymore.

File: packages/acceptance-tests/pkg-tests-specs/sources/pnpapi.test.ts [hunk 4]
```diff
@@ -49,8 +45,8 @@ describe(`Plug'n'Play API`, () => {
           [`no-deps`]: `1.0.0`,
         },
       }, async ({path, run, source}) => {
-        await xfs.mkdirpPromise(`${path}/workspace`);
-        await xfs.writeJsonPromise(`${path}/workspace/package.json`, {
+        await xfs.mkdirpPromise(`${path}/workspace` as PortablePath);
+        await xfs.writeJsonPromise(`${path}/workspace/package.json as PortablePath` as PortablePath, {
```
[the remark this answers is not in the document]
Comment (paul-soporan) [reply packages/acceptance-tests/pkg-tests-specs/sources/pnpapi.test.ts]: 😄 Thanks for fixing!

**Comment by arcanis** [comment]
Tests almost pass, but need to be adjusted for Windows support 👍

File: packages/yarnpkg-pnp/sources/loader/makeApi.ts [hunk 5]
```diff
@@ -851,11 +853,12 @@ export function makeApi(runtimeState: RuntimeState, opts: MakeApiOptions): PnpAp
 
   function resolveRequest(request: PortablePath, issuer: PortablePath | null, {considerBuiltins, extensions, conditions, require}: ResolveRequestOptions = {}): PortablePath | null {
     const unqualifiedPath = resolveToUnqualified(request, issuer, {considerBuiltins});
-
     if (unqualifiedPath === null)
       return null;
 
-    const remappedPath = resolveUnqualifiedExport(request, unqualifiedPath, {conditions, require});
+    const remappedPath = !considerBuiltins || !builtinModules.has(request)
```
Comment (paul-soporan) [inline packages/yarnpkg-pnp/sources/loader/makeApi.ts:859]: Shouldn't the same logic also be applied when `request` is `pnpapi`?

Comment (arcanis) [reply packages/yarnpkg-pnp/sources/loader/makeApi.ts:859]: It can't be `pnpapi`, it has already been resolved to a file path by this point (inside `resolveToUnqualified`).

Comment (paul-soporan) [reply packages/yarnpkg-pnp/sources/loader/makeApi.ts:859]: Oh, indeed. 🤦‍♂️

**Review by merceyz — APPROVED** [review]

File: packages/yarnpkg-pnp/sources/loader/makeApi.ts (hunk 2 above)
Comment (SimenB) [reply packages/yarnpkg-pnp/sources/loader/makeApi.ts]: Seems unfortunate that node's defaults should be forced on everybody using `exports` field. For a module that isn't supported to run in Node, it should just fail to resolve or something (or falling back to `default`) not pick up `node`.

There's already a switch in the library for `browser` condition so it's not consistently enforced - why not just fully let the caller provide their own conditions (barring possibly `default` - that seems reasonable to always add)? I.e. remove the two `allows.add` calls (and thus the options leaving only `conditions`). Or probably add `node` if no `conditions` is passed, but otherwise add nothing.

---

I was gonna open up an issue over in `resolve.exports`, but since I came over this PR with this comment before I got to it, I sneakily commented here instead 😀 Happy to open up an issue over there instead

Comment (SimenB) [reply packages/yarnpkg-pnp/sources/loader/makeApi.ts]: As another datapoint, TypeScript could not have used `resolve.exports` to implement support for `types` condition with their official example: https://devblogs.microsoft.com/typescript/announcing-typescript-4-5-beta/#packagejson-exports-imports-and-self-referencing

I think `resolve.exports` should be implementation independent (maybe a `resolve.exports-node` or just a `import { node } from 'resolve.exports'` makes sense, but the base export shouldn't put these limitations there)
